File: src/fragment/StringFragment.ts
import { Text } from 'pixi.js';
import Fragment, { FragmentType, register } from './Fragment';
import { StructEndecBuilder, PrimitiveEndecs } from 'KEndec';

export default class StringFragment extends Fragment {
    string: string;

    constructor(string: string) {
        super();

        this.string = string;
    }

    override toString(): string {
        return this.string;
    }

    override type(): FragmentType<StringFragment> {
        return SLOT;
    }
}

const SLOT = register(
    'string',
    0xaabb77,
    StructEndecBuilder.of1(
        PrimitiveEndecs.STRING.fieldOf('value', (fragment: StringFragment) => fragment.string),
        (str) => new StringFragment(str)
    )
);

File: src/jsMain/kotlin/me/maplesyrum/tricksterstudio/spell/fragment/StringFragment.kt
package me.maplesyrum.tricksterstudio.spell.fragment

import tree.maple.kendec.PrimitiveEndecs
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder

data class StringFragment(val value: String) : Fragment() {
    override fun type(): FragmentType<*> {
        return FragmentType.STRING
    }

    override fun toString(): String {
        return value
    }

    companion object {
        val ENDEC: StructEndec<StringFragment> = StructEndecBuilder.of(
            PrimitiveEndecs.STRING.fieldOf("value", StringFragment::value),
            { value: String -> StringFragment(value) }
        )
    }
}
